src/gleam/result.gleam

```diff
@@ -411,3 +411,36 @@ pub fn replace_error(result: Result(a, e1), error: e2) -> Result(a, e2) {
 pub fn values(results: List(Result(a, e))) -> List(a) {
   list.filter_map(results, fn(r) { r })
 }
+
+/// Run a callback function if the given bool is `True`, otherwise return a
+/// default error value.
+///
+/// This function is suitable for `use` expressions.
+///
+/// ## Examples
+///
+/// ```gleam
+/// > let name = "Kamaka"
+/// > use <- guard(name != "", or: "Missing name")
+/// > "Hello, " <> name
+/// Ok("Hello, Kamaka")
+/// ```
+///
+/// ```gleam
+/// > let name = ""
+/// > use <- guard(name != "", or: "Missing name")
+/// > "Hello, " <> name
+/// Error("Missing name")
+/// ```
+///
+///
+pub fn guard(
```

Comment:
On my first read I was a bit surprised to see that this doesn't take in some `Result(a, e)` as the value to test.

I can see why it is the way it is, but I think there's a possibility of misunderstanding here – particularly gonna be confusing if someone tries to use this in a pipe like:
```gleam
Ok("louis")
|> result.guard(fn (name) { name != "louis" }, NoLouis)
|> ...
```
which I think would be a reasonable interpretation of the function if they weren't paying attention to the types.

Could we have best of both worlds?
```gleam
pub fn guard(
  value: Result(a, e),
  against requirement: fn(a) -> Bool,
  or error: e,
  then consequence: fn(a) -> Result(b, e),
) -> Result(b, e) {
  use a <- result.try(value)
  case requirement(a) 
    True -> consequence(a)
    False -> Error(error)
  }
}
```
```gleam
let name = Ok("Kamaka")
use valid_name <- guard(name, fn(name) { name != "" }, MissingName)
Ok("Hello, " <> name)
```

Guess this gets a bit convoluted now. I appreciate this is straying quite far from what was here initially too...